docs: add preliminary knowledge about collektive and the paradigm (#548)
Collektive/collektive: site/docs/tutorials/basic-usage.mdx

    environment.set<Int>("maxNeighborID", maxValue)

    return maxValue
}

```
The library function `neighboringViaExchange(local)` is a **aggregate operator** and is used to observe the value of an expression [local] across neighboring nodes.

#### Example

```kotlin
val field = neighboring(0)
```

The returned field has as its local value the value passed as input (0 in this example).

**Note**: the scenario in which the functions are applied is modeled using the [Alchemist simulator](https://alchemistsimulator.github.io/), enabling a precise and controlled representation of network dynamics.

<details>
  <summary>Click to view images</summary>
  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-first-example-1.png" style={{ maxWidth: '70%' }} />
  </div>

  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-first-example-2.png" style={{ maxWidth: '70%' }} />
  </div>

  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-first-example-3.png" style={{ maxWidth: '70%' }} />
  </div>

  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-first-example-4.png" style={{ maxWidth: '70%' }} />

>   </div>

Reviewer: When displaying example images, provide a brief description of what can be seen in them.

>   In the third step of the code from the first example, the computed values are assigned to molecules. These molecules can be inspected within the data of a selected device, thereby simplifying the debugging process and enabling a more effective analysis of the system's behavior.

Reviewer: Molecules and relative concentrations are strictly related to the Alchemist Simulator. 
In the Collektive first-steps documentation, try to avoid them as much as possible because the tool should potentially run also outside the Alchemist environment.

    // Step 2: Find the maximum value among neighbors (including self)
    val maxValue = neighborValues.max(base = localId)

    // Step 3: Assign the result to a molecule
    environment.set<Boolean>("isMaxLocalID", localId == maxValue)
    environment.set<Int>("maxNeighborID", maxValue)

    return maxValue
}

fun Aggregate<Int>.searchMaxNetworkIDValue(environment: EnvironmentVariables): Int {
    environment.set<Int>("localID", localId)

    when (environment.get<Boolean>("isMaxLocalID")) {
        true -> searchMaxNeighborIDValue(environment)
        false -> environment.get<Int>("maxNeighborID")
    }

    // Step 1: Exchange the maxNeighborID with neighbors and obtain a field of values
    val neighborValues = neighboringViaExchange(local = environment.get<Int>("maxNeighborID"))

    // Step 2: Find the maximum value among neighbors (including self)
    val maxValue = neighborValues.max(base = environment.get<Int>("maxNeighborID"))

    // Step 3: Assign the result to a molecule
    environment.set<Boolean>("isMaxID", localId == maxValue)
    environment.set<Int>("maxNetworkID", maxValue)
    
    return maxValue

> }

Reviewer: Use the suggestions above to fix this code.

```
<details>
  <summary>Click to view images</summary>
  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-second-example-1.png" style={{ maxWidth: '70%' }} />
  </div>

  The network remains unchanged from the first example, except for the interconnection radius between devices. A poorly connected network would result in the presence of multiple global maxima within the network.

  <div style={{ textAlign: 'center' }}>

>     <img src="/img/basic-usage-second-example-2.png" style={{ maxWidth: '70%' }} />

Reviewer: The network is too dense and displaying the connections in such a network is hard to understand. 
It would be better to reduce its density first and then show the connections.

  </div>

>   This second example represents a highly simplified demonstration of leader election using the Collektive library.

Reviewer: Explain better what can be observed by the image.

---
title: Basic Usage
---

# Simple examples

## First example

A preliminary code example demonstrates how devices, through mutual communication, are able to identify the maximum ID among the neighboring nodes within their local network area (neighborhood):

```kotlin
/**
 * First example - Tutorial:
 * 1. Identify the maximum value among the neighboring nodes.
 * 2. Assign a distinct color to the nodes with the identified maximum values.
*/

fun Aggregate<Int>.maxID(environment: EnvironmentVariables) = maxNeighborID(environment)

fun Aggregate<Int>.maxNeighborID(environment: EnvironmentVariables): Int {
    // Step 1: Exchange the localId with neighbors and obtain a field of values
    val neighborValues = neighboring(local = localId)

    // Step 2: Find the maximum value among neighbors (including self)

>     val maxValue = neighborValues.max(base = localId)

Reviewer: ok for the division into step to clarify the meaning of the operation, but I would add in another equivalent example the online code, in order to highlight the ease of use and the short syntax.

  
  The resulting configuration after executing the code shown above is as follows:
  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-second-example-2.png" style={{ maxWidth: '70%' }} />
  </div>

  There are four interconnected subnets, and the number of interconnected subnets is determined by the number of highlighted nodes. The nodes within each of these subnets communicate the local maximum value identified within them, thereby enabling the identification of the global maximum value of the entire network.
  <div style={{ textAlign: 'center' }}>
    <img src="/img/basic-usage-second-example-1.png" style={{ maxWidth: '70%' }} />
  </div>
  For the sake of visual clarity, the network is presented without the connection edges between the nodes. 
  
  Additionally, it should be noted that fewer nodes have been included compared to the previous example.
</details>

> This second example represents a highly simplified demonstration of leader election using the Collektive library.

Reviewer: ```suggestion
This second example illustrates a variant of leader election implemented with Collektive.
```

Specifically, in this first two examples, the use of aggregated operators and operations on the neighborhood field is analyzed.

## Third example

> In this final example, we extend the previous exercise with the goal of determining the diameter in relation to a source corresponding to the global maximum. To make the exercise more interesting, we reduce the number of nodes and the distance at which two nodes are connected, in order to obtain a larger number of diameter values. For each time step in the simulation, the average of these values will be calculated. 

Reviewer: explain better what is the objective of this example, also it is not clear what you mean with the "the diameter in relation to a source etc", provide a more meaningful definition.
